File: automaguard-java/src/main/java/io/automaguard/PolicyEngine.java
package io.automaguard;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;

/**
 * AutomaGuard policy engine.
 *
 * <p>Load a compiled {@code .aegisc} policy <em>once</em> at agent startup,
 * then call {@link #evaluate(String, Map)} on each agent event.  Evaluation
 * is synchronous and typically completes in under 1 ms.  The engine is
 * thread-safe — the native Rust side is protected by a {@code Mutex}.</p>
 *
 * <p>Implements {@link AutoCloseable}; use in a try-with-resources block to
 * ensure the native engine handle is freed:</p>
 *
 * <pre>{@code
 * try (PolicyEngine engine = PolicyEngine.fromFile("guard.aegisc")) {
 *     PolicyResult result = engine.evaluate("tool_call", Map.of(
 *         "tool_name", "send_email",
 *         "arguments", Map.of("to", "user@example.com")
 *     ));
 *     if (result.isDenied()) {
 *         throw new EnforcementException(result);
 *     }
 * }
 * }</pre>
 */
public final class PolicyEngine implements AutoCloseable {

    private static final ObjectMapper MAPPER = PolicyResult.MAPPER;

    private final NativeBridge bridge;
    private final long handle;
    private final AtomicLong eventCount = new AtomicLong(0);
    private volatile boolean closed = false;

    /** Package-private constructor for production and testing. */
    PolicyEngine(NativeBridge bridge, long handle) {
        this.bridge = bridge;
        this.handle = handle;
    }

    // ── Factory methods ───────────────────────────────────────────────────────

    /**
     * Load a policy engine from a compiled {@code .aegisc} file.
     *
     * @param path absolute or relative path to the {@code .aegisc} file
     * @return a new {@code PolicyEngine}
     * @throws IOException if the file cannot be found or is not a valid policy
     */
    public static PolicyEngine fromFile(String path) throws IOException {
        if (path == null || path.isBlank()) {
            throw new IllegalArgumentException("path must not be null or blank");
        }
        NativeBridge bridge = new JniNativeBridge();
        long handle = bridge.fromFile(path);
        return new PolicyEngine(bridge, handle);
    }

    /**
     * Load a policy engine from raw {@code .aegisc} bytes (e.g. embedded as
     * a class-path resource).
     *
     * @param data compiled policy bytes
     * @return a new {@code PolicyEngine}
     * @throws IOException if the bytes cannot be parsed as a valid policy
     */
    public static PolicyEngine fromBytes(byte[] data) throws IOException {
        if (data == null) {
            throw new IllegalArgumentException("data must not be null");
        }
        NativeBridge bridge = new JniNativeBridge();
        long handle = bridge.fromBytes(data);
        return new PolicyEngine(bridge, handle);
    }

    // ── Properties ────────────────────────────────────────────────────────────

    /**
     * The policy name declared in the {@code .aegisc} file
     * ({@code policy <Name> \{…\}}).
     */
    public String getPolicyName() {
        ensureOpen();
        return bridge.policyName(handle);
    }

    /** Total events evaluated since creation or the last {@link #reset()}. */
    public long getEventCount() { return eventCount.get(); }

    // ── Core API ──────────────────────────────────────────────────────────────

    /**
     * Evaluate a single agent event against the loaded policy.
     *
     * @param eventType event type string (e.g. {@code "tool_call"},
     *                  {@code "data_access"}, {@code "external_request"})
     * @param fields    event field map; values must be JSON-serialisable.
     *                  Pass {@code null} or an empty map for events with no fields.
     * @return the {@link PolicyResult} containing verdict, triggered rules, and
     *         any violations
     */
    public PolicyResult evaluate(String eventType, Map<String, Object> fields) {
        if (eventType == null || eventType.isBlank()) {
            throw new IllegalArgumentException("eventType must not be null or blank");
        }
        ensureOpen();

        String fieldsJson = null;
        if (fields != null && !fields.isEmpty()) {
            try {
                fieldsJson = MAPPER.writeValueAsString(fields);
            } catch (Exception e) {
                throw new RuntimeException("Failed to serialise event fields to JSON", e);
            }
        }

        String resultJson = bridge.evaluate(handle, eventType, fieldsJson);
        eventCount.incrementAndGet();
        return PolicyResult.fromJson(resultJson);
    }

    /**
     * Evaluate an event with no additional fields.
     *
     * @param eventType event type string
     * @return the {@link PolicyResult}
     */
    public PolicyResult evaluate(String eventType) {
        return evaluate(eventType, null);
    }

    /**
     * Reset all state machines and rate-limit counters to their initial states.
     *
     * <p>Call this at the start of a new agent session to isolate per-session
     * temporal invariants and rate limits.</p>
     */
    public void reset() {
        ensureOpen();
        bridge.reset(handle);
        eventCount.set(0);
    }

    // ── AutoCloseable ─────────────────────────────────────────────────────────

    /**
     * Release the native engine handle.
     *
     * <p>After closing, any call to this instance will throw
     * {@link IllegalStateException}.  Closing is idempotent.</p>
     */
    @Override
    public void close() {
        if (!closed) {
            closed = true;
            bridge.free(handle);
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void ensureOpen() {
        if (closed) {
            throw new IllegalStateException("PolicyEngine has been closed");
        }
    }
}

File: automaguard-dotnet/src/AutomaGuard/PolicyEngine.cs
using System.Text.Json;
using System.Threading;

namespace AutomaGuard;

/// <summary>
/// AutomaGuard policy engine.
///
/// Load a compiled <c>.aegisc</c> policy <b>once</b> at agent startup, then
/// call <see cref="Evaluate"/> on each agent event. Evaluation is synchronous
/// and typically completes in under 1 ms. The engine is thread-safe — the
/// native Rust side is protected by a <c>Mutex</c>.
///
/// <example>
/// <code>
/// using var engine = PolicyEngine.FromFile("guard.aegisc");
///
/// var result = engine.Evaluate("tool_call", new Dictionary&lt;string, object?&gt;
/// {
///     ["tool_name"] = "send_email",
///     ["arguments"] = new { to = "user@example.com" },
/// });
///
/// if (result.Verdict == Verdict.Deny)
///     throw new EnforcementException(result);
/// </code>
/// </example>
/// </summary>
public sealed class PolicyEngine : IDisposable
{
    private readonly INativeEngine _native;
    private long _eventCount;
    private bool _disposed;

    /// <summary>Internal constructor — accepts any <see cref="INativeEngine"/> (real or mock).</summary>
    internal PolicyEngine(INativeEngine native)
    {
        _native = native;
    }

    // ── Factory methods ───────────────────────────────────────────────────────

    /// <summary>
    /// Load a policy engine from a compiled <c>.aegisc</c> file.
    /// </summary>
    /// <param name="path">Absolute or relative path to the <c>.aegisc</c> file.</param>
    /// <exception cref="InvalidOperationException">
    /// The file does not exist or is not a valid compiled policy.
    /// </exception>
    public static PolicyEngine FromFile(string path)
    {
        ArgumentException.ThrowIfNullOrEmpty(path);
        return new PolicyEngine(new RealNativeEngine(path));
    }

    /// <summary>
    /// Load a policy engine from raw <c>.aegisc</c> bytes (e.g. embedded as
    /// an assembly resource).
    /// </summary>
    /// <param name="data">Compiled policy bytes.</param>
    /// <exception cref="InvalidOperationException">
    /// The buffer is not a valid compiled policy.
    /// </exception>
    public static PolicyEngine FromBytes(byte[] data)
    {
        ArgumentNullException.ThrowIfNull(data);
        return new PolicyEngine(new RealNativeEngine(data));
    }

    // ── Properties ────────────────────────────────────────────────────────────

    /// <summary>The policy name declared in the <c>.aegisc</c> file (<c>policy Name { }</c>).</summary>
    public string PolicyName
    {
        get
        {
            ThrowIfDisposed();
            return _native.PolicyName;
        }
    }

    /// <summary>Total events evaluated since creation or the last <see cref="Reset"/>.</summary>
    public long EventCount => Interlocked.Read(ref _eventCount);

    // ── Core API ──────────────────────────────────────────────────────────────

    /// <summary>
    /// Evaluate a single agent event against the loaded policy.
    /// </summary>
    /// <param name="eventType">
    /// Event type string (e.g. <c>"tool_call"</c>, <c>"data_access"</c>).
    /// </param>
    /// <param name="fields">
    /// Arbitrary event fields. Pass <see langword="null"/> or omit for events
    /// with no fields. Values are serialised to JSON before being passed to
    /// the native engine.
    /// </param>
    /// <returns>
    /// A <see cref="PolicyResult"/> containing the verdict, triggered rules,
    /// and any violations.
    /// </returns>
    public PolicyResult Evaluate(
        string eventType,
        IReadOnlyDictionary<string, object?>? fields = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(eventType);
        ThrowIfDisposed();

        string? fieldsJson = fields is null or { Count: 0 }
            ? null
            : JsonSerializer.Serialize(fields, JsonOptions.Default);

        PolicyResult result = _native.Evaluate(eventType, fieldsJson);
        Interlocked.Increment(ref _eventCount);
        return result;
    }

    /// <summary>
    /// Set a persistent context value accessible in policy expressions as
    /// <c>context.&lt;key&gt;</c>. Values accumulate across events for the
    /// lifetime of this engine instance.
    /// </summary>
    /// <param name="key">Context key.</param>
    /// <param name="value">
    /// Value — any JSON-serialisable type (string, number, bool, object).
    /// </param>
    public void SetContext(string key, object? value)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        ThrowIfDisposed();
        string valueJson = JsonSerializer.Serialize(value, JsonOptions.Default);
        _native.SetContext(key, valueJson);
    }

    /// <summary>
    /// Set a policy configuration value accessible in policy expressions as
    /// <c>config.&lt;key&gt;</c>.
    /// </summary>
    /// <param name="key">Config key.</param>
    /// <param name="value">Value — any JSON-serialisable type.</param>
    public void SetConfig(string key, object? value)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        ThrowIfDisposed();
        string valueJson = JsonSerializer.Serialize(value, JsonOptions.Default);
        _native.SetConfig(key, valueJson);
    }

    /// <summary>
    /// Reset all state machines and rate-limit counters to their initial states.
    ///
    /// Call this at the start of a new agent session to isolate per-session
    /// temporal invariants and rate limits.
    /// </summary>
    public void Reset()
    {
        ThrowIfDisposed();
        _native.Reset();
        Interlocked.Exchange(ref _eventCount, 0);
    }

    /// <summary>Return a snapshot of the engine's current operational state.</summary>
    public EngineStatus Status()
    {
        ThrowIfDisposed();
        return _native.Status();
    }

    // ── IDisposable ───────────────────────────────────────────────────────────

    /// <summary>
    /// Release the native engine handle.
    ///
    /// After disposal any call to this instance will throw
    /// <see cref="ObjectDisposedException"/>.
    /// </summary>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _native.Dispose();
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void ThrowIfDisposed()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
    }
}
